src/GraphCoarsenBuilder.cpp

```diff
@@ -190,17 +243,258 @@ void AssembleMBuilder::AddTraceAcross(int row, int col, double value)
+                comp_EF_E_.push_back(RTDP(P_EF, local_fine_weight, P_E));
+                // also store transpose, or just have it implicitly?
+            }
+        }
+    }
+    components_built_ = true;
+}
+
+/// this shares a lot of code with BuildComponents, but I'm not sure it makes
+/// sense to combine them in any way.
+std::unique_ptr<mfem::SparseMatrix> CoefficientMBuilder::GetCoarseM(
+    const mfem::Vector& fineMdiag,
+    const mfem::SparseMatrix& Pedges, const mfem::SparseMatrix& face_cdof)
+{
+    // build the components, if necessary
+    if (!components_built_)
+    {
+        BuildComponents(fineMdiag, Pedges, face_cdof);
+    }
+
+    const int num_aggs = topology_.Agg_face_.Height();
+    const int num_faces = topology_.Agg_face_.Width();
+
+    // ---
+    // assemble from components...
+    // ---
+    auto CoarseM = make_unique<mfem::SparseMatrix>(
+                       total_num_traces_ + ncoarse_vertexdofs_ - num_aggs,
+                       total_num_traces_ + ncoarse_vertexdofs_ - num_aggs);
+
+    // F_F block, the P_F^T M_F P_F pieces
+    mfem::Array<int> neighbor_aggs;
+    mfem::Array<int> coarse_face_dofs;
+    for (int face = 0; face < num_faces; ++face)
+    {
+        double face_weight;
+        GetTableRow(topology_.face_Agg_, face, neighbor_aggs);
+        MFEM_ASSERT(neighbor_aggs.Size() <= 2, "Face has three or more aggregates!");
+        if (neighbor_aggs.Size() == 1)
+        {
+            face_weight = agg_weights_[neighbor_aggs[0]];
+        }
+        else
+        {
+            face_weight = 2.0 / (1.0 / agg_weights_[neighbor_aggs[0]] +
+                                 1.0 / agg_weights_[neighbor_aggs[1]]);
+        }
+        GetCoarseFaceDofs(face_cdof, face, coarse_face_dofs);
+        AddScaledSubMatrix(*CoarseM, coarse_face_dofs, coarse_face_dofs,
+                           comp_F_F_[face], face_weight);
+    }
+
+    // the EF_EF block
+    // for (pairs of *faces* that share an *aggregate*)
+    mfem::Array<int> local_faces;
+    mfem::Array<int> coarse_face_dofs_prime;
+    int counter = 0;
+    for (int agg = 0; agg < num_aggs; ++agg)
+    {
+        double agg_weight = agg_weights_[agg];
+        GetTableRow(topology_.Agg_face_, agg, local_faces);
+        for (int f = 0; f < local_faces.Size(); ++f)
+        {
+            int face = local_faces[f];
+            GetCoarseFaceDofs(face_cdof, face, coarse_face_dofs);
+            for (int fprime = f; fprime < local_faces.Size(); ++fprime)
+            {
+                int faceprime = local_faces[fprime];
+                GetCoarseFaceDofs(face_cdof, faceprime, coarse_face_dofs_prime);
+                AddScaledSubMatrix(*CoarseM, coarse_face_dofs,
+                                   coarse_face_dofs_prime, comp_EF_EF_[counter],
+                                   agg_weight);
+                if (f != fprime)
+                {
+                    mfem::DenseMatrix EFEFtrans(comp_EF_EF_[counter], 't');
```

Comment:
The explicit copy and transpose can be avoided by inserting the submatrix transposed, but I'm not sure that it is worth the hassle.

Reply:
That's not a bad idea, but it would involve nontrivial modifications to `AddScaledSubMatrix`. At this stage I don't think it's worth the hassle.